### PR TITLE
feat!: implement ADR 21:  Subspace specific custom fee tokens

### PR DESCRIPTION
**Pull request by dadamu**
## Description

This PR implements ADR-21: Subspace specific custom fee tokens feature.

Closes: [DCD-312](https://desmoslabs.atlassian.net/browse/DCD-312)

~~Depends-On: #1135~~
~~Depends-On: #1139~~

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/desmos-labs/desmos/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://docs.cosmos.network/v0.44/building-modules/intro.html)
- [x] included the necessary unit and integration [tests](https://github.com/desmos-labs/desmos/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

[DCD-312]: https://forbole.atlassian.net/browse/DCD-312?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by dadamu** [comment]
@RiccardoM @manu0466 Rebased

File: x/subspaces/ante/tx_checker.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/ante"
+	antetypes "github.com/desmos-labs/desmos/v5/x/subspaces/ante/types"
+)
+
+// CheckTxFeeWithSubspaceMinPrices returns the tx checker that including the subspace allowed tokens into minimum prices list
+func CheckTxFeeWithSubspaceMinPrices(txFeeChecker ante.TxFeeChecker, sk antetypes.SubspacesKeeper) ante.TxFeeChecker {
+	return func(ctx sdk.Context, tx sdk.Tx) (sdk.Coins, int64, error) {
+		subspaceID, isSubspaceTx := GetTxSubspaceID(tx)
+		if !isSubspaceTx {
```
Comment (dadamu) [inline x/subspaces/ante/tx_checker.go:13]: @RiccardoM
Do you think we should exclude subspace management messages ( `MsgEditSubspace`, `MsgAddUserToUserGroup` and etc.), forcing admins mange subspace by using `DSM`?

File: proto/desmos/subspaces/v3/msgs.proto [hunk 2]
```diff
@@ -492,3 +500,33 @@ message MsgSetUserPermissions {
 // MsgSetUserPermissionsResponse defines the Msg/SetPermissionsResponse
 // response type
 message MsgSetUserPermissionsResponse {}
+
+// --------------------------------------------------------------------------------------------------------------------
+
+// MsgUpdateSubspaceFeeTokens represents the message to be used to update a subspace fee tokens by governance
```
Comment (RiccardoM) [inline proto/desmos/subspaces/v3/msgs.proto:506]:
```suggestion
// MsgUpdateSubspaceFeeTokens represents the message to be used to update the accepted fee tokens
// inside a given subspace, using an on-chain governance proposal
```

File: proto/desmos/subspaces/v3/msgs.proto [hunk 3]
```diff
@@ -492,3 +500,33 @@ message MsgSetUserPermissions {
...
+message MsgUpdateSubspaceFeeTokens {
+  option (cosmos.msg.v1.signer) = "authority";
+
+  // Id of the subspace where the list of allowed fee tokens will be updated
+  uint64 subspace_id = 1 [
+    (gogoproto.customname) = "SubspaceID",
+    (gogoproto.moretags) = "yaml:\"subspace_id\""
+  ];
+    
+  // List of the allowed tokens to be fee token inside the subspace along with their default minimum prices
```
Comment (RiccardoM) [inline proto/desmos/subspaces/v3/msgs.proto:518]:
```suggestion
  // List of tokens allowed to be fee tokens inside the given subspace, represented as their gas prices
```

File: x/subspaces/ante/tx_checker.go (hunk 1 above)
Comment (RiccardoM) [reply x/subspaces/ante/tx_checker.go:13]: Yes, I believe we should *not* allow those messages to be paid using non-DSM tokens.

File: x/subspaces/ante/tx_checker.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+		}
+
+		subspace, found := sk.GetSubspace(ctx, subspaceID)
+		if !found {
+			return txFeeChecker(ctx, tx)
+		}
+
+		newMinPrices := MergeMinPrices(ctx.MinGasPrices(), sdk.NewDecCoinsFromCoins(subspace.AllowedFeeTokens...))
+		newCtx := ctx.WithMinGasPrices(newMinPrices)
+		return txFeeChecker(newCtx, tx)
```
Comment (RiccardoM) [inline x/subspaces/ante/tx_checker.go:24]: I'm not super familiar with on-chain gas prices, but I thought they were treated using the `and` logic. So if I have a min gas price of `0.01udsm,0.01ustar` then to properly broadcast my transaction I have to pay fees in *both* tokens, and not *either one* of the two. Do you have any link for this to be checked out that we can paste here to remember this in the future?

File: x/subspaces/client/cli/cli_test.go [hunk 5]
```diff
@@ -195,7 +195,7 @@ func (s *IntegrationTestSuite) TestCmdQuerySubspace() {
 					"cosmos1s0he0z3g92zwsxdj83h0ky9w463sx7gq9mqtgn",
 					"cosmos1s0he0z3g92zwsxdj83h0ky9w463sx7gq9mqtgn",
 					time.Date(2020, 1, 1, 12, 00, 00, 000, time.UTC),
-				),
+				).SetAllowedFeeTokens(sdk.Coins{}),
```
Comment (RiccardoM) [inline x/subspaces/client/cli/cli_test.go:198]: Why is this needed?

File: x/subspaces/keeper/msg_server.go [hunk 6]
```diff
@@ -831,3 +832,44 @@ func (k msgServer) SetUserPermissions(goCtx context.Context, msg *types.MsgSetUs
 
 	return &types.MsgSetUserPermissionsResponse{}, nil
 }
+
+// UpdateSubspaceFeeTokens defines a rpc method for MsgUpdateSubspaceFeeTokens
+func (k msgServer) UpdateSubspaceFeeTokens(goCtx context.Context, msg *types.MsgUpdateSubspaceFeeTokens) (*types.MsgUpdateSubspaceFeeTokensResponse, error) {
+	authority := k.authority
+	if authority != msg.Authority {
+		return nil, errors.Wrapf(govtypes.ErrInvalidSigner, "invalid authority; expected %s, got %s", authority, msg.Authority)
+	}
```
Comment (RiccardoM) [inline x/subspaces/keeper/msg_server.go:841]: No need to delcare a variable here if we're not going to use it anywhere outside this check
```suggestion
	if msg.Authority != k.authority {
		return nil, errors.Wrapf(govtypes.ErrInvalidSigner, "invalid authority; expected %s, got %s", k.authority, msg.Authority)
	}
```

File: x/subspaces/keeper/msg_server_test.go [hunk 7]
```diff
@@ -3059,3 +3060,125 @@ func (suite *KeeperTestSuite) TestMsgServer_SetUserPermissions() {
...
+			},
+			msg: types.NewMsgUpdateSubspaceFeeTokens(
+				1,
+				sdk.Coins{{Denom: "minttoken", Amount: sdk.NewInt(-10)}},
+				"cosmos1x5pjlvufs4znnhhkwe8v4tw3kz30f3lxgwza53",
+			),
+			shouldErr: true,
+		},
+		{
+			name: "subspace are updated correctly",
```
Comment (RiccardoM) [inline x/subspaces/keeper/msg_server_test.go:3112]:
```suggestion
			name: "subspace is updated correctly",
```

File: x/subspaces/types/models.go [hunk 8]
```diff
@@ -84,9 +84,22 @@ func (sub Subspace) Validate() error {
...
+		return fmt.Errorf("invalid allowed fee tokens: %s", sub.AllowedFeeTokens)
+	}
+
 	return nil
 }
 
+// Set allowed fee tokens without validating it.
+// Before storing the updated subspace, a validation with Validate() should
+// be performed.
+func (sub Subspace) SetAllowedFeeTokens(feeTokens sdk.Coins) Subspace {
```
Comment (RiccardoM) [inline x/subspaces/types/models.go:98]: I'm not a big fan of this approach. Instead I think it would be better to: 

1. Add another param inside the `NewSubspace` function to specify the fee tokens when creating the subspace. This is going to be useful when exporting/importing the genesis as well as inside tests. 
2. Rename this method `Update` as we have done with the `Post` and `Attachment` type, to keep everything consistent.

Also, I'm doubting whether `AllowedFeeTokens` is the proper name. I think `AdditionalFeeTokens` might be better instead. The name `AllowedFeeTokens` might make people think that if I specify `0.01ustar` then the `star` token is going to be then only one allowed, while the `udsm` should always be allowed anyway. Using `AdditionalFeeTokens` might make everything more easy to understand.

File: x/subspaces/ante/tx_checker.go (hunk 4 above)
Comment (dadamu) [reply x/subspaces/ante/tx_checker.go:24]: They are treated as `or`,  so one of tokens reaches the min fees can broadcast properly, say, to send the message with 100 gas via the node with `0.01udsm, 0.01ustar` min gas prices, `1udsm` or `1ustar` can pass the check.

Here is the reference:
https://github.com/cosmos/cosmos-sdk/blob/release/v0.47.x/x/auth/ante/validator_tx_fee.go#L38

File: x/subspaces/types/models.go (hunk 8 above)
Comment (dadamu) [reply x/subspaces/types/models.go:98]: I see, I use the `SetAllowedFeeTokens` since adding another param inside the `NewSubspace` will break hundreds of existing test codes, but yeah, it is better to have an additional field.

strongly agree with `AdditionalFeeTokens` is better.

Comment (dadamu) [reply x/subspaces/types/models.go:98]: Finally, using `sed` to update the new field quickly:
```bash
#!/bin/bash

dir="./x"

find "$dir" -type f -name "*.go" | while read -r file; do
    sed -i -E ':a;N;$!ba;s/NewSubspace\([^)]*blockTime,/&\nnil,/g' "$file"
    echo "$file"
done
```

File: x/subspaces/client/cli/cli_test.go (hunk 5 above)
Comment (dadamu) [reply x/subspaces/client/cli/cli_test.go:198]: It is added by mistake, removed.

Comment (dadamu) [reply x/subspaces/client/cli/cli_test.go:198]: It is added to avoid from returning `nil` which actually expected to be `empty` slice.
It can be solved by adding new parameter to `NewSubspace`.

File: x/subspaces/ante/tx_checker.go (hunk 1 above)
Comment (RiccardoM) [reply x/subspaces/ante/tx_checker.go]: I think it's best to do the opposite: adding a `ManageSubspaceMsg` interface that identifies only the one that should *not* be included. This is faster and easier to maintain for future features: it's going to be less likely to add new management operations rather than social operations

File: x/subspaces/types/models.go (hunk 8 above)
Comment (RiccardoM) [reply x/subspaces/types/models.go]: I honestly don't know of any other tool that can easily update the code. I would have done it by hand :sweat: Great job using `sed` :100:

File: x/subspaces/ante/tx_checker.go (hunk 1 above)
Comment (dadamu) [reply x/subspaces/ante/tx_checker.go:26]: Great idea, Implemented `ManageSubspaceMsg`, instead.

**Review by RiccardoM — APPROVED** [review]
Great job :100:
